pkg/metabase/grant.go

```diff
@@ -0,0 +1,397 @@
+	}
+
+	exists, _ := memberExists(mbGroupMembers, s[1])
+	if exists {
+		log.Info("member already exists")
+		return
+	}
+
+	if err := m.client.AddPermissionGroupMember(ctx, mbMetadata.PermissionGroupID, s[1]); err != nil {
+		log.WithError(err).WithField("user", s).
+			WithField("group", mbMetadata.PermissionGroupID).
+			Warn("Unable to sync user")
+	}
+}
+
+func (m *Metabase) createRestricted(ctx context.Context, ds *models.Dataset) error {
+	groupID, err := m.client.CreatePermissionGroup(ctx, ds.Name)
+	if err != nil {
+		return err
+	}
+
+	aadGroupID, err := m.client.CreatePermissionGroup(ctx, ds.Name+" (aad)")
+	if err != nil {
+		return err
+	}
+
+	// Hack/workaround necessary due to how metabase has implemented saml group sync. When removing a saml group mapping to a
+	// metabase permission group, the users in the saml group will remain members of the permission group.
+	// Adding a dummy mapping ensures that users are evicted from the permission group when an actual group mapping is removed.
+	// See https://github.com/metabase/metabase/issues/26079
+	if err := m.client.UpdateGroupMapping(ctx, "non-existant-aad-group", aadGroupID, GroupMappingOperationAdd); err != nil {
+		return err
+	}
+
+	colID, err := m.client.CreateCollectionWithAccess(ctx, []int{groupID, aadGroupID}, ds.Name)
+	if err != nil {
+		return err
+	}
+
+	key, email, err := m.createServiceAccount(ds)
+	if err != nil {
+		return err
+	}
+
+	err = m.create(ctx, dsWrapper{
+		Dataset:            ds,
+		Key:                string(key),
+		Email:              email,
+		MetabaseGroupID:    groupID,
+		MetabaseAADGroupID: aadGroupID,
+		CollectionID:       colID,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *Metabase) create(ctx context.Context, ds dsWrapper) error {
+	datasource, err := m.repo.GetBigqueryDatasource(ctx, ds.Dataset.ID)
+	if err != nil {
+		return err
+	}
+
+	err = m.accessMgr.Grant(ctx, datasource.ProjectID, datasource.Dataset, datasource.Table, "serviceAccount:"+ds.Email)
+	if err != nil {
+		return err
+	}
+
+	dp, err := m.repo.GetDataproduct(ctx, ds.Dataset.DataproductID)
+	if err != nil {
+		return err
+	}
+
+	dbID, err := m.client.CreateDatabase(ctx, dp.Owner.Group, ds.Dataset.Name, ds.Key, ds.Email, &datasource)
+	if err != nil {
+		return err
+	}
+
+	err = m.repo.CreateMetabaseMetadata(ctx, models.MetabaseMetadata{
+		DatasetID:            ds.Dataset.ID,
+		DatabaseID:           dbID,
+		PermissionGroupID:    ds.MetabaseGroupID,
+		AADPermissionGroupID: ds.MetabaseAADGroupID,
+		CollectionID:         ds.CollectionID,
+		SAEmail:              ds.Email,
+	})
+	if err != nil {
+		return err
+	}
+
+	m.waitForDatabase(ctx, dbID, datasource.Table)
+
+	if ds.MetabaseGroupID > 0 || ds.MetabaseAADGroupID > 0 {
+		err := m.client.RestrictAccessToDatabase(ctx, []int{ds.MetabaseGroupID, ds.MetabaseAADGroupID}, dbID)
+		if err != nil {
+			return err
+		}
+	}
+
+	if err := m.HideOtherTables(ctx, dbID, datasource.Table); err != nil {
```

Comment:
It seems if error happens in this step, all tables may expose to the granted group. Can we do this before granting access?

Reply:
If it fails it will not cause the granted group to gain access to all tables in the bigquery dataset. The only table the group will be granted access to is the one we grant on https://github.com/navikt/nada-backend/blob/hard_delete/pkg/metabase/grant.go#L266. The reason we are hiding tables is that all tables in the bigquery dataset will appear in metabase even though the group members does not have permission to query them. So it's more a clean up to avoid confusion. If we move the hiding of tables prior to granting access we would have to move the creation of the database as well, which in turn depends on grant.

Reply:
I see! Thank you